README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->
![](man/figures/pltesim_logo_annimation.gif)

> R package to simulate **P**robabilistic **L**ong-**T**erm **E**ffects
> in models with temporal dependence

Christopher Gandrud and Laron K. Williams

Version: 1.0 [![CRAN
Version](http://www.r-pkg.org/badges/version/pltesim)](https://CRAN.R-project.org/package=pltesim)
[![Build
Status](https://travis-ci.org/christophergandrud/pltesim.svg?branch=master)](https://travis-ci.org/christophergandrud/pltesim)
![CRAN Monthly
Downloads](http://cranlogs.r-pkg.org/badges/last-month/pltesim) ![CRAN
Total Downloads](http://cranlogs.r-pkg.org/badges/grand-total/pltesim)

About
=====

**pltesim** implements [Williams's
(2016)](http://pan.oxfordjournals.org/content/24/2/243) method for
simulating probabilistic long-term effects in models with temporal
dependence.

It is built on the
[coreSim](https://github.com/christophergandrud/coreSim) package.

Process
=======

To find and show probabilistic long-term effects in models with temporal
dependence with **pltesim**:

1.  Estimate the coefficients. Currently **pltesim** works with binary
    outcome models, e.g. logit, so use `glm` from the default
    R installation.

2.  Create a data frame with your counterfactual. This should have a row
    with the fitted counterfactual values and columns with names
    matching those in your fitted model. All variables without values
    will be treated as 0 in the counterfactual.

3.  Simulate the long-term effects with `plte_builder`.

4.  Plot the results with `plte_plot`.

Examples
========

These examples replicate Figure 1 in [Williams
(2016)](http://pan.oxfordjournals.org/content/24/2/243). First estimate
your model. You may need to use `btscs` to generate spells for the
binary dependent variable.

    library(pltesim)
    library(ggplot2)

    data('negative_year')

    # BTSCS set the data
    neg_set <- btscs(df = negative_year, event = 'y', t_var = 'year',
                     cs_unit = 'group', pad_ts = FALSE)

    # Estimate the model
    m1 <- glm(y ~ x + spell_time + I(spell_time^2) + I(spell_time^3),
              family = binomial(link = 'logit'),
              data = neg_set)

Then fit the counterfactual:

    counterfactual <- data.frame(x = 0.5)

Now simulate and plot long-term effects for a variety of scenarios using
`plte_builder` and `plte_plot`. `plte_builder` takes as its input the
fitted model object with the estimated coefficients (`obj`), an
identification of the basic time period variable (`obj_tvar`), the
counterfactual (`cf`), how long the counterfactual persists
(`cf_duration`, it is `permanent` by default), and the time period
points over which to simulate the effects.

Note that by default the predicted probabilities from logistic
regression models are found. You can specify a custom quantity of
interest function with the `FUN` argument.

In this first example the counterfactual is persistent throughout the
entire time span:

    # Permanent
    sim1 <- plte_builder(obj = m1, obj_tvar = 'spell_time',
                         cf = counterfactual, t_points = c(13, 25))

    plte_plot(sim1) + ggtitle('Permanent')

![](man/figures/perm-1.png)

Note that the numbers next to each simulation point indicate the time
since the last event. You can choose to not show these numbers by
setting `t_labels = FALSE` in the `plte_plot` call.

In the next example, the effect only lasts for one time period:

    # One-time
    sim2 <- plte_builder(obj = m1, obj_tvar = 'spell_time', cf_duration = 'one-time',
                         cf = counterfactual, t_points = c(13, 25))

    plte_plot(sim2) + ggtitle('One-time')

![](man/figures/onetime-1.png)

We can also have the counterfactual effect last for short periods of
time and simulate the effect if another event occurs:

    # Temporary
    sim3 <- plte_builder(obj = m1, obj_tvar = 'spell_time', cf_duration = 4,
                         cf = counterfactual, t_points = c(13, 25))

    plte_plot(sim3) + ggtitle('Temporary')

![](man/figures/temp-multievent-1.png)

    # Multiple events, permanent counter factual
    sim4 <- plte_builder(obj = m1, obj_tvar = 'spell_time',
                         cf = counterfactual, t_points = c(13, 20, 25))

    plte_plot(sim4) + ggtitle('Permanent, Multiple Events')

![](man/figures/temp-multievent-2.png)

By default the baseline scenario has all covariate values fitted at 0.
You can supply a custom baseline scenario in the second row of the
counterfactual (`cf`) data frame. For example:

    # Custom baseline scenario
    counterfactual_baseline <- data.frame(x = c(1, 0.5))

    sim5 <- plte_builder(obj = m1, obj_tvar = 'spell_time', cf_duration = 4,
                         cf = counterfactual_baseline, t_points = c(13, 25))

    plte_plot(sim5) + ggtitle('Temporary, Custom Baseline')

![](man/figures/custom-baseline-1.png)


NEWS
# Version 1.0

- !!Breaking change!!--`btscs` now starts the spell time counter from 1,
rather than 0. The returned variable is now called `spell_time` rather than
`spell`. The former has a more intuitive interpretation.

- `btscs` checks that all supplied variables are in `df`. Error otherwise.

- Includes a modified version of the `negative` data set called
`negative_year`. This enables more intuitive examples.


# Version 0.1.2

- Linetype for `plte_plot` now matches Williams (2016), e.g. solid lines for
Y = 1.

- Minor documentation improvements.

# Version 0.1.1

- `plte_builder` finds polynomial splines using the B-spline basis. Note: apart
from the degree of the piecewise polynomial, the default behaviour for `bs`
is assumed.

- Added ability to supply a custom baseline scenario to `plte_builder via
the `cf` argument.

# Version 0.1.0

- Minimal working version.


tests/testthat.R
library(testthat)
library(pltesim)
library(car)
library(survival)

test_check("pltesim")


tests/testthat/test.R
data('negative')

# Test btscs -------------------------------------------------------------------
neg_df <- btscs(negative, 'y', 'tim', 'group')

test_that('btscs output validity', {
          expect_equal(nrow(neg_df), 1000)
          expect_equal(ncol(neg_df), 7)
          expect_equal(sum(neg_df[['spell_time']]), 10327)
})


R/btscs.R
#' Generate spells for binary variables
#'
#' @param df a data frame
#' @param event character string naming the binary variable identifying the
#' event. Note \code{1} must signify events and \code{0} non-events.
#' @param t_var character string with the name of the time variable.
#' @param cs_unit character string with the name of the cross-sectional unit.
#' @param pad_ts logical indicating whether or not to fill in the time-series
#' if panels are unbalanced.
#'
#' @return The original (\code{df}) data frame with an additional
#' \code{spell_time} value identifying the number of observed periods in the spell,
#' i.e. time points since the last period.
#'
#' @examples
#' data('negative')
#'
#' neg_set <- btscs(df = negative, event = 'y', t_var = 'tim', cs_unit = 'group')
#'
#' @source This function is a port of Dave Armstrong's \code{btscs} function
#' from:
#'
#' Dave Armstrong (2015). DAMisc: Dave Armstrong's Miscellaneous Functions.
#' R package version 1.3. \url{https://CRAN.R-project.org/package=DAMisc}.
#'
#' It was ported largely to reduce the dependencies needed for the examples.
#' There are also internal improvements, largely to handle single period spells
#' and to start the spell time counter from 1.
#'
#' David's package implemented the Stata function from:
#'
#' Beck, N.. J. Katz and R. Tucker. 1998. "Beyond Ordinary Logit: Taking Time
#' Seriously in Binary-Time-Series-Cross-Section Models". American Journal of
#' Political Science 42(4): 1260-1288.
#'
#' @export

btscs <- function (df, event, t_var, cs_unit, pad_ts = FALSE)
{
    if (!is.data.frame(df)) stop('df must be a data frame.', call. = FALSE)

    df$orig_order <- 1:nrow(df)
    df <- df[order(df[[cs_unit]], df[[t_var]]), ]
    spells <- function(x) {
        tmp <- rep(0, length(x))
        runcount <- 0
        if (length(x) == 1) tmp <- runcount
        else
            for (j in 2:length(x)) {
                if (x[j] == 0 & x[(j - 1)] == 0) {
                    tmp[j] <- runcount <- runcount + 1
                }
                if (x[j] != 0 & x[(j - 1)] == 0) {
                    tmp[j] <- runcount + 1
                    runcount <- 0
                }
                if (x[j] == 0 & x[(j - 1)] != 0) {
                    tmp[j] <- runcount <- 0
                }
            }
        tmp
    }
    sp <- split(df, df[[cs_unit]])
    if (pad_ts) {
        sp <- lapply(sp, function(x) x[match(seq(min(x[[t_var]], na.rm = TRUE),
                                                 max(x[[t_var]], na.rm = TRUE)),
                                             x[[t_var]]),
                                       ])
        for (i in seq_along(sp)) {
            if (any(is.na(sp[[i]][[event]]))) {
                sp[[i]][[event]][which(is.na(sp[[i]][[event]]))] <- 1
            }
            if (any(is.na(sp[[i]][[t_var]]))) {
                sp[[i]][[t_var]] <- seq(min(sp[[i]][[t_var]], na.rm = TRUE),
                                       max(sp[[i]][[t_var]], na.rm = TRUE))
            }
            if (any(is.na(sp[[i]][[cs_unit]]))) {
                sp[[i]][[cs_unit]][which(is.na(sp[[i]][[cs_unit]]))] <-
                    mean(sp[[i]][[cs_unit]], na.rm = TRUE)
            }
        }
    }
    sp <- lapply(seq_along(sp), function(x) {
        cbind(sp[[x]], data.frame(spell_time = spells(sp[[x]][[event]])),
              row.names = NULL)
    })
    df <- do.call(rbind, sp)
    if (!pad_ts) {
        if (any(is.na(df$orig_order))) {
            df <- df[-which(is.na(df$orig_order)), ]
        }
        df <- df[df$orig_order, ]
    }
    else {
        df <- df[order(df[[cs_unit]], df[[t_var]]), ]
    }
    df$orig_order <- NULL
    df$spell_time <- df$spell_time + 1
    return(df)
}


R/plot.plte.R
#' Plot objects created by plte_builder
#'
#' @param obj a \code{plte} class object created by \code{\link{plte_builder}}.
#' @param t_labels logical whether or not to include time labels for each point.
#'
#' @return A \code{gg} ggplot2 object that can be modified using the
#' \code{+} in combination with other ggplot2 functions.
#'
#' @examples
#' data('negative')
#'
#' # BTSCS set the data
#' neg_set <- btscs(df = negative, event = 'y', t_var = 'tim',
#'                  cs_unit = 'group', pad_ts = FALSE)
#'                  # Create temporal dependence variables
#' neg_set$t <- neg_set$spell + 1
#'
#' m1 <- glm(y ~ x + t + I(t^2) + I(t^3),
#'           family = binomial(link = 'logit'),
#'           data = neg_set)
#'
#' counterfactual <- data.frame(x = 0.5)
#'
#' sim1 <- plte_builder(obj = m1, obj_tvar = 't',
#'                      cf = counterfactual, t_points = c(13, 25),
#'                      cf_duration = 4, ci = 99)
#'
#' # With time point labels
#' plte_plot(sim1)
#'
#' # Without time point labels
#' plte_plot(sim1, t_labels = FALSE)
#'
#'
#' @import ggplot2
#'
#' @export

plte_plot <- function(obj, t_labels = TRUE)
{
    scenario_name <- scenario_time <- qi_median <- qi_min <- qi_max <- t__ <- NULL

    if (!inherits(obj, 'plte'))
        stop('obj must be plte class created by plte_builder.', call. = FALSE)

    ot <- attr(obj, 'obj_tvar')
    names(obj)[names(obj) == ot] <- 't__'
    obj[1, 't__'] <- ''

    sims <- subset(obj, !(scenario_name == 'counterfactual' & scenario_time < 3))

    sims$scenario_name <- factor(sims$scenario_name,
                                 levels = c('counterfactual', 'baseline'),
                                 labels = c('Y = 1', 'Y = 0'))

    p <- ggplot(sims, aes(x = scenario_time, y = qi_median,
                          group = scenario_name)) +
        geom_point(position = position_dodge(width = 0.3)) +
       geom_linerange(aes(ymin = qi_min, ymax = qi_max,
                            linetype = scenario_name),
                        position = position_dodge(width = 0.3)) +
        scale_linetype_discrete(name = 'Counterfactual') +
        scale_x_continuous(breaks = 1:3, labels = c('Baseline', 't', 't + 1')) +
        xlab('') + ylab('Pr(Y = 1)\n') +
        theme_bw()

    if (t_labels)
        p <- p + geom_text(aes(label = t__), alpha = 0.5, nudge_x = 0.3)

    p
}

R/plte_builder.R
#' Create simulations for long-term effects in models with temporal dependence
#'
#' @param obj a fitted model object.
#' @param obj_tvar character string specifying the name of the base time variable
#' in \code{obj}.
#' @param cf a data frame with the first row containing the counterfactual.
#' An optional second row could be supplied with values for the baseline
#' scenario. If not supplied then all values are set to zero for the baseline.
#' Columns should have names that match variables in \code{obj} and contain
#' fitted values to find quantities of interest for. Note, this should not
#' include your time variable as this is handled using \code{t_points}.
#' @param cf_duration a character string or numeric specifying the
#' counterfactual's duration. If \code{'permanent'} then the
#' counterfacutal lasts for the full time span in \code{t_points}. If
#' \code{'one-time'} then the counterfactual only lasts for one period.
#' If \code{cf_duration} is numeric then the number specifies the
#' number of time increments at which the counterfactual resets.
#' @param t_points a numeric vector with a minimum length of 2 and a maximum
#' lentgh of 3. The first and last values should be the time starting and ending
#' points for the simulatinos. The (optional) middle value can specify a point
#' between the first and last time points where a subsequent event occurs.
#' @param FUN a function for finding a quantity of interest from the linear
#' systematic component. See \code{\link{qi_builder}}. The default is a function
#' for finding the predicted probability from a logistic regression model.
#' @param nsim number of simulations to draw.
#' @param ci the proportion of the central interval of the simulations to
#' return. Must be in (0, 1] or equivalently (0, 100].
#'
#' @examples
#' data('negative')
#'
#' # BTSCS set the data
#' neg_set <- btscs(df = negative, event = 'y', t_var = 'tim',
#'                 cs_unit = 'group', pad_ts = FALSE)
#'
#' # Create temporal dependence variable
#' neg_set$t <- neg_set$spell + 1
#'
#' m1 <- glm(y ~ x + t + I(t^2) + I(t^3),
#'           family = binomial(link = 'logit'), data = neg_set)
#'
#' # Create fitted counterfactual
#' counterfactual <- data.frame(x = 0.5)
#'
#' # Permanent counterfactual, one event
#' sim1 <- plte_builder(obj = m1, obj_tvar = 't',
#'                      cf = counterfactual, t_points = c(13, 25))
#'
#' # Multiple events
#' sim2 <- plte_builder(obj = m1, obj_tvar = 't',
#'                      cf = counterfactual, t_points = c(13, 18, 25))
#'
#' # One-time counterfactual
#' sim3 <- plte_builder(obj = m1, obj_tvar = 't',
#'                      cf = counterfactual, t_points = c(13, 25),
#'                      cf_duration = 'one-time')
#'
#' # Temporary (4 period counterfactual)
#' sim4 <- plte_builder(obj = m1, obj_tvar = 't',
#'                      cf = counterfactual, t_points = c(13, 25),
#'                      cf_duration = 4)
#'
#' # Custom baseline scenario
#' # Note: the second row is the custom baseline
#' counterfactual_baseline <- data.frame(x = c(1, 0.5))
#'
#' sim5 <- plte_builder(obj = m1, obj_tvar = 't', cf_duration = 4,
#'                      cf = counterfactual_baseline, t_points = c(13, 25))
#'
#' # Time splines
#' library(splines)
#' m2 <- glm(y ~ x + bs(t, degree = 3), family = binomial(link = 'logit'),
#'           data = neg_set)
#'
#' sim6 <- plte_builder(obj = m2, obj_tvar = 't', cf_duration = 4,
#'                      cf = counterfactual, t_points = c(13, 25))
#'
#' @source
#' Williams, Laron K. 2016. "Long-Term Effects in Models with Temporal
#' Dependence". Political Analysis: 24(2): 243-262.
#'
#' @importFrom coreSim qi_builder
#'
#' @export

plte_builder <- function(obj, obj_tvar,
                         cf, cf_duration = 'permanent',
                         t_points, FUN = logistic_prob_FUN,
                         ci = 0.95, nsim = 1000)
{
    # Create scenarios to simulate ---------------------------------------------
    if (missing(obj_tvar))
        stop('obj_tvar must be specified.', call. = FALSE)
    if (!(obj_tvar %in% bs_stripper(names(obj$coefficients))))
        stop('Cannot find obj_tvar in the fitted model object.', call. = FALSE)
    if (obj_tvar %in% colnames(cf)) {
        message('It is not necessary to include the time variable in cf.',
                call. = FALSE)
        cf[, obj_tvar] <- NULL
    }

    if (length(t_points) < 2)
        stop('Must specify at least two t_range values.', call. = FALSE)
    if (length(t_points) > 3)
        stop('t_points can only include 2 or three time points.', call. = FALSE)

    t_start = t_points[1]
    t_end = t_points[length(t_points)]
    t_no_events <- c(t_start, t_start:t_end)
    t_change <- c(rep(t_start, 2), 0:(t_end - t_start - 1))

    if (length(t_points) == 3) {
        temp <- t_no_events
        for (u in seq_along(temp)) {
            if (t_points[2] <= temp[u]) temp[u] <- temp[u] - t_points[2]
            if (temp[u] < t_change[u]) t_change[u] <- temp[u]
        }
    }

    nrow_cf <- nrow(cf)
    if (!(nrow_cf %in% 1:2))
        stop('cf must have only one or two rows.', call. = FALSE)

    # Create baseline scenario if not supplied
    names_cf <- names(cf)
    if (nrow_cf == 1) {
        make_zero <- function(x) x * 0
        baseline <- data.frame(apply(cf, 1, make_zero))
    }
    else if (nrow_cf == 2) {
        baseline <- data.frame(cf[2, ])
        cf <- data.frame(cf[1, ])
        names(cf) <- names_cf
    }
    names(baseline) <- names_cf

    # Find counterfactuals over time
    if (length(cf_duration) != 1)
        stop('cf_duration can only have one value', call. = FALSE)

    npost_base <- (t_end - t_start) + 1

    if (cf_duration == tolower('permanent')) {
        post_base_cf <- df_repeat(cf, npost_base)
        cf <- rbind(baseline, post_base_cf)
    }
    else if (cf_duration == tolower('one-time')) {
        cf <- rbind(baseline, cf, df_repeat(baseline, npost_base - 1))
    }
    else if (is.numeric(cf_duration)) { # temporary
        cf <- rbind(df_repeat(baseline, 1), df_repeat(cf, cf_duration),
                    df_repeat(baseline, npost_base - cf_duration))
    }

    scenarios <- list(baseline = data.frame(cf, t_no_events),
                      counterfactual = data.frame(cf, t_change))

    # Find quantities of interest ----------------------------------------------
    sims <- data.frame()
    for (i in 1:length(scenarios)) {
        temp <- scenarios[[i]]
        names(temp) <- c(names(temp)[-ncol(temp)], obj_tvar)

        temp_sims <- data.frame()
        for (u in 1:nrow(temp)) {
            temp_1 <- temp[u, ]
            temp_1 <- qi_builder(obj = obj, newdata = temp_1, FUN = FUN,
                                    ci = ci, nsim = nsim, slim = TRUE,
                                    original_order = TRUE)
            temp_sims <- rbind(temp_sims, temp_1)
        }
        temp_sims$scenario_name <- names(scenarios)[i]
        temp_sims$scenario_time <- 1:nrow(temp_sims)
        sims <- rbind(sims, temp_sims)
    }

    ncs <- ncol(sims)
    sims <- cbind(sims[, c((ncs-1):ncs)], sims[, c(1:(ncs-2))])

    class(sims) <- c('data.frame', 'plte')
    attr(sims, 'obj_tvar') <- obj_tvar
    return(sims)

}
